### PR TITLE
Generalize FETools::Compositing::multiply_dof_numbers()

### PR DESCRIPTION
**Pull request by peterrum**
needed for #12647

### CODE REVIEW

File: include/deal.II/fe/fe_base.h [hunk 1]
```diff
@@ -1023,24 +1023,34 @@ internal::GenericDoFsPerObject::generate(const FiniteElementData<dim> &fe)
...
+
+  result.first_object_index_on_face.resize(3);
   for (unsigned int face_no : reference_cell.face_indices())
     {
       result.first_object_index_on_face[0].emplace_back(0);
 
-      if (dim >= 2)
-        result.first_object_index_on_face[1].emplace_back(
-          fe.get_first_face_line_index(face_no));
+      // if (dim >= 2)
```
Comment (drwells) [inline include/deal.II/fe/fe_base.h:1047]:
```suggestion
```

File: include/deal.II/fe/fe_base.h [hunk 2]
```diff
@@ -1023,24 +1023,34 @@ internal::GenericDoFsPerObject::generate(const FiniteElementData<dim> &fe)
...
+      result.first_object_index_on_face[1].emplace_back(
+        fe.get_first_face_line_index(face_no));
 
-      if (dim == 3)
-        result.first_object_index_on_face[2].emplace_back(
-          fe.get_first_face_quad_index(face_no));
+      // if (dim == 3)
```
Comment (drwells) [inline include/deal.II/fe/fe_base.h:1051]:
```suggestion
```

File: include/deal.II/fe/fe_base.h [hunk 3]
```diff
@@ -1023,24 +1023,34 @@ internal::GenericDoFsPerObject::generate(const FiniteElementData<dim> &fe)
...
+      result.first_object_index_on_face[2].emplace_back(
+        fe.get_first_face_quad_index(face_no));
```
Comment (drwells) [inline include/deal.II/fe/fe_base.h:1053]: I don't understand what's going on on these two lines - could you add a comment?

File: include/deal.II/fe/fe_tools.templates.h [hunk 4]
```diff
@@ -103,23 +98,62 @@ namespace FETools
...
-            multiplied_dofs_per_line +=
-              fes[i]->n_dofs_per_line() * multiplicities[i];
-            multiplied_dofs_per_quad +=
-              fes[i]->n_dofs_per_quad(0) * multiplicities[i];
-            multiplied_dofs_per_hex +=
-              fes[i]->n_dofs_per_hex() * multiplicities[i];
+          dpos_in[i] =
+            dealii::internal::GenericDoFsPerObject::generate(*fes[i]);
+
+      const auto initialize =
```
Comment (drwells) [inline include/deal.II/fe/fe_tools.templates.h:110]: This part is also pretty hard to understand compared to the old version. Could you add a sentence or two here explaining how we set up each field in `dpo`?

**Review by drwells — CHANGES_REQUESTED** [review]

File: include/deal.II/fe/fe_tools.templates.h (hunk 4 above)
Comment (peterrum) [reply include/deal.II/fe/fe_tools.templates.h]: I have added some comments.

File: include/deal.II/fe/fe_base.h (hunk 3 above)
Comment (peterrum) [reply include/deal.II/fe/fe_base.h:1051]: These lines fill `first_object_index_on_face`. Entry 0 are vertices, 1 are lines, and 2 are "quads". What comments are you suggesting?

Comment (drwells) [reply include/deal.II/fe/fe_base.h:1051]: Its clearer now that I read it again - I think its ok as-is.

**Review by drwells — APPROVED** [review]

**Pull request merged**